server/application_api: TestDrainSqlStats_partialOutage failed

server/application_api.TestDrainSqlStats_partialOutage [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/21338248?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/21338248?buildTab=artifacts#/) on master @ [7c25eaa9935f9d881adf62a5388c044598991839](https://github.com/cockroachdb/cockroach/commits/7c25eaa9935f9d881adf62a5388c044598991839):


```
=== RUN   TestDrainSqlStats_partialOutage
    test_log_scope.go:172: test logs captured to: /artifacts/tmp/_tmp/cd6614104b60b7c80a95bb0194031a15/logTestDrainSqlStats_partialOutage1627002850
    test_log_scope.go:82: use -show-logs to present logs inline
    test_server_shim.go:168: automatically injected a shared process virtual cluster under test; see comment at top of test_server_shim.go for details.
    sql_stats_test.go:2089: 
        	Error Trace:	pkg/server/application_api/sql_stats_test.go:2089
        	            				pkg/server/application_api/sql_stats_test.go:1630
        	Error:      	Not equal: 
        	            	expected: 91
        	            	actual  : 89
        	Test:       	TestDrainSqlStats_partialOutage
    panic.go:694: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/cd6614104b60b7c80a95bb0194031a15/logTestDrainSqlStats_partialOutage1627002850
--- FAIL: TestDrainSqlStats_partialOutage (7.83s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/obs-prs
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestDrainSqlStats_partialOutage.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-63643

server/application_api.TestDrainSqlStats_partialOutage [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/21339107?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/21339107?buildTab=artifacts#/) on master @ [1a34c229dfa74568ba33af51109dc8c86b297264](https://github.com/cockroachdb/cockroach/commits/1a34c229dfa74568ba33af51109dc8c86b297264):


```
=== RUN   TestDrainSqlStats_partialOutage
    test_log_scope.go:172: test logs captured to: /artifacts/tmp/_tmp/7d791e7a1abfe4198e60f6d6453f914e/logTestDrainSqlStats_partialOutage2943523868
    test_log_scope.go:82: use -show-logs to present logs inline
    test_server_shim.go:168: automatically injected a shared process virtual cluster under test; see comment at top of test_server_shim.go for details.
    test_server_shim.go:637: DRPC is enabled (via testserver factory defaults)
    sql_stats_test.go:2211: 
        	Error Trace:	pkg/server/application_api/sql_stats_test.go:2211
        	            				pkg/server/application_api/sql_stats_test.go:1752
        	Error:      	Not equal: 
        	            	expected: 96
        	            	actual  : 89
        	Test:       	TestDrainSqlStats_partialOutage
    panic.go:694: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/7d791e7a1abfe4198e60f6d6453f914e/logTestDrainSqlStats_partialOutage2943523868
--- FAIL: TestDrainSqlStats_partialOutage (7.16s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestDrainSqlStats_partialOutage.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>

--- a/pkg/server/sql_stats.go
+++ b/pkg/server/sql_stats.go
@@ -28,6 +28,7 @@ import (
 // different aggregation windows are not combined.
 type DrainSqlStatsRespBuilder struct {
 	stmtFingerprintCount map[appstatspb.StmtFingerprintID]struct{}
+	txnFingerprintCount  map[appstatspb.TransactionFingerprintID]struct{}
 	stmtMap              map[stmtDrainKey]*appstatspb.CollectedStatementStatistics
 	txnMap               map[txnDrainKey]*appstatspb.CollectedTransactionStatistics
 }
@@ -49,6 +50,7 @@ type txnDrainKey struct {
 func NewDrainSqlStatsRespBuilder() *DrainSqlStatsRespBuilder {
 	return &DrainSqlStatsRespBuilder{
 		stmtFingerprintCount: make(map[appstatspb.StmtFingerprintID]struct{}),
+		txnFingerprintCount:  make(map[appstatspb.TransactionFingerprintID]struct{}),
 		stmtMap:              make(map[stmtDrainKey]*appstatspb.CollectedStatementStatistics),
 		txnMap:               make(map[txnDrainKey]*appstatspb.CollectedTransactionStatistics),
 	}
@@ -84,11 +86,15 @@ func (b *DrainSqlStatsRespBuilder) AddTxnStats(
 		} else {
 			existingTx.Stats.Add(&txn.Stats)
 		}
+
+		if _, ok := b.txnFingerprintCount[txn.TransactionFingerprintID]; !ok {
+			b.txnFingerprintCount[txn.TransactionFingerprintID] = struct{}{}
+		}
 	}
 }
 
 func (b *DrainSqlStatsRespBuilder) Build() *serverpb.DrainStatsResponse {
-	fingerprintCount := len(b.stmtFingerprintCount) + len(b.txnMap)
+	fingerprintCount := len(b.stmtFingerprintCount) + len(b.txnFingerprintCount)
 	response := &serverpb.DrainStatsResponse{
 		Statements:       make([]*appstatspb.CollectedStatementStatistics, 0, len(b.stmtMap)),
 		Transactions:     make([]*appstatspb.CollectedTransactionStatistics, 0, len(b.txnMap)),

